Derive P2PKH, P2WPKH  and P2SH-P2WPKH address from BIP32 key

Sjors opened the pull request:
Introduces function to derive a legacy (P2PKH) or wrapped SegWit (P2SH-P2WPKH) address from a BIP32 key:

```c
int wally_bip32_key_to_address(const struct ext_key *hdkey, uint32_t flags,
                               uint32_t version, char **output)
```

Implements #102 except for native SegWit, which I'll add a separate function for.

I did not add `c++` headers in `wally.hpp`.

I added tests in Python, but not for Javascript, Swig Python and Swig Java.

Sjors commented:
@glslang thanks for your review. Rebased and hopefully addresses all comments.

1. I ran `./tools/uncrustify`, but it only fixed one indentation. Is there another linter that you're running, or is this all manual? I renamed `pubKey(Hash)` to `pub_key(_hash)`, `redeemScript(Hash)` to `redeem_script(_hash)`. I removed braces around  `return WALLY_EINVAL`.

2. Fixed

3. Fixed (2x)

glslang commented:
unfortunately is manual currently. many thanks for the changes.

Sjors commented:
Added native SegWit in a separate commit.

jkauffman1 commented on src/address.c:
```diff
@@ -0,0 +1,81 @@
+
+        ret = wally_base58_from_bytes(address, sizeof(address), BASE58_FLAG_CHECKSUM, output);
+    } else {
+        BUILD_ASSERT(flags == WALLY_ADDRESS_TYPE_P2SH_P2WPKH);
```
Should this be a runtime check?

Sjors replied:
This should never happen, because of the above `Catch known incorrect combinations` checks. I don't know if the compiler is smart enough to understand that though. If not, then I should make it a runtime check (but I couldn't get `assert` to work).

glslang replied:
the comment has more to do with the fact that this isn't a compile time expression and will cause a sizeof evaluatiion of a VLA effectively. I think it needs to be converted to an if expression or removed.

Sjors replied:
This goes way beyond my C comprehension. I'll just drop the `BUILD_ASSERT` then.

jkauffman1 commented on src/address.c:
```diff
@@ -0,0 +1,81 @@
+
+    if (flags == WALLY_ADDRESS_TYPE_P2PKH) {
+        // pub_key_hash = ripemd160(sha256(pubkey)
+        address[0] = (unsigned char) version & 0xff;
```
Could possibly be pulled out of the if

Sjors commented:
Addressed @jkauffman1's suggestion to move `ret = wally_base58_from_bytes(...` out of the if statement. Not sure what to do about the `BUILD_ASSERT`.

Added testnet vectors.

Sjors commented:
Travis failure looks random to me (`No available formula with the name "swig@3.04"`)

glslang commented on src/address.c:
```diff
@@ -0,0 +1,77 @@
+        return WALLY_EINVAL;
+    }
+
+    unsigned char address[HASH160_LEN + 1];
```
style: move to before "int ret" on line 12

glslang commented on src/address.c:
```diff
@@ -0,0 +1,77 @@
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x00) ||
+        (flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0xC4) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x6F)
+        ) {
```
nit: remove { to conform to code style

Sjors commented:
Done and rebased.

jkauffman1 commented on src/address.c:
```diff
@@ -0,0 +1,73 @@
+    if ((flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0x05) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x00) ||
+        (flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0xC4) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x6F))
```
Maybe it's worth defining these constants (0, 5, 111, 196) especially as they are being used in the related PR: https://github.com/ElementsProject/libwally-core/pull/112

Sjors replied:
Done

greenaddress commented:
utack eb5257cb314dab9e9fe17bb360cdf03b88f17e97
